internal/controllers/nmc_reconciler.go

```diff
@@ -321,14 +321,23 @@ func (h *nmcReconcilerHelperImpl) ProcessModuleSpec(
 	}
 
 	if pod == nil {
+		// new module is introduced, need to load it
 		if status == nil {
 			logger.Info("Missing status; creating loader Pod")
 			return h.pm.CreateLoaderPod(ctx, nmcObj, spec)
 		}
 
+		/* configuration changed for module: if spec status contain the same kernel,
+		unload the kernel module, otherwise - load kernel modules, since the pod
+		is not running, the module cannot be loaded using the old kernel configuration
+		*/
 		if !reflect.DeepEqual(spec.Config, status.Config) {
-			logger.Info("Outdated config in status; creating unloader Pod")
-			return h.pm.CreateUnloaderPod(ctx, nmcObj, status)
+			if spec.Config.KernelVersion == status.Config.KernelVersion {
+				logger.Info("Outdated config in status; creating unloader Pod")
+				return h.pm.CreateUnloaderPod(ctx, nmcObj, status)
+			}
+			logger.Info("Outdated config in status and kernels differ, probably due to upgrade; creating loader Pod")
+			return h.pm.CreateLoaderPod(ctx, nmcObj, spec)
```

Comment:
Why do we need to create the loader pod here?

In this case, the node's reboot ready state is newer than the config status so the loader pod will be created in line 348 anyway isn't it?

Reply:
I thought about that, but eventually went with this solution for a number of reasons:
1) this makes the code more readable, since the conditions are stated clearly
2) makes unit tests more readable and easier to check

Reply:
I would agree if this was a separate case. But IMO it falling exactly in the "node rebooting" option.

After all:
1. the node has rebooted
2. It has rebooted after the last update on the NMC status
3. If an unloaderPod was required, then it has already been done.

Seems like a pretty good fit to me.

Reply:
That condition was added just for node reboot scenario. I agree it will work, but it will be much harder to understand and especially debug such cases.

Reply:
We discussed offline, we will refactor this function in a later PR to be in a switch-case flow which will make it clearer - in this case, it is indeed clearer to have it separate by conditions and each condition to contain all it actions.

LGTM